runtime/common/src/paras_registrar.rs

```diff
@@ -0,0 +1,674 @@
+		/// `gensis_head` and `validation_code` are used to initalize the parathread's state.
+		#[weight = 0]
+		fn register_parathread(
+			origin,
+			id: ParaId,
+			genesis_head: HeadData,
+			validation_code: ValidationCode,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!Paras::contains_key(id), Error::<T>::ParaAlreadyExists);
+
+			let outgoing = <paras::Module<T>>::outgoing_paras();
+
+			ensure!(outgoing.binary_search(&id).is_err(), Error::<T>::ParaAlreadyExists);
+
+			<T as Trait>::Currency::reserve(&who, T::ParathreadDeposit::get())?;
+			<Debtors<T>>::insert(id, who);
+
+			Paras::insert(id, false);
+
+			let genesis = ParaGenesisArgs {
+				genesis_head,
+				validation_code,
+				parachain: false,
+			};
+
+			<paras::Module<T>>::schedule_para_initialize(id, genesis);
+
+			Ok(())
+		}
+
+		/// Deregister a parathread and retreive the deposit.
+		///
+		/// Must be sent from a `Parachain` origin which is currently a parathread.
+		///
+		/// Ensure that before calling this that any funds you want emptied from the parathread's
+		/// account is moved out; after this it will be impossible to retreive them (without
+		/// governance intervention).
+		#[weight = 0]
+		fn deregister_parathread(origin) -> DispatchResult {
+			let id = ensure_parachain(<T as Trait>::Origin::from(origin))?;
+
+			let is_parachain = Paras::take(id).ok_or(Error::<T>::InvalidChainId)?;
+
+			ensure!(!is_parachain, Error::<T>::InvalidThreadId);
+
+			let debtor = <Debtors<T>>::take(id);
+			let _ = <T as Trait>::Currency::unreserve(&debtor, T::ParathreadDeposit::get());
+
+			<paras::Module<T>>::schedule_para_cleanup(id);
+
+			Ok(())
+		}
+
+		/// Swap a parachain with another parachain or parathread. The origin must be a `Parachain`.
+		/// The swap will happen only if there is already an opposite swap pending. If there is not,
+		/// the swap will be stored in the pending swaps map, ready for a later confirmatory swap.
+		///
+		/// The `ParaId`s remain mapped to the same head data and code so external code can rely on
+		/// `ParaId` to be a long-term identifier of a notional "parachain". However, their
+		/// scheduling info (i.e. whether they're a parathread or parachain), auction information
+		/// and the auction deposit are switched.
+		#[weight = 0]
+		fn swap(origin, other: ParaId) {
+			let id = ensure_parachain(<T as Trait>::Origin::from(origin))?;
+
+			if PendingSwap::get(other) == Some(id) {
+				// Remove intention to swap.
+				PendingSwap::remove(other);
+
+				Paras::mutate(id, |i|
+					Paras::mutate(other, |j|
+						sp_std::mem::swap(i, j)
+					)
+				);
+
+				<Debtors<T>>::mutate(id, |i|
+					<Debtors<T>>::mutate(other, |j|
+						sp_std::mem::swap(i, j)
+					)
+				);
+			} else {
+				PendingSwap::insert(id, other);
+			}
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Register a parachain with given code. Must be called by root.
+	/// Fails if given ID is already used.
+	pub fn register_parachain(
```

Comment:
I note this doesn't set `Debtors`. The previous logic used something called `SwapAux` because deposits for parachains are in the `slots` module. I would prefer to have all deposits in one place, and this includes XCMP/HRMP channel deposits. So at some point we should have a `deposits` module where we can implement swapping over. I'm fine with the code committed as is for now, though.